dawn/src/dawn/Optimizer/PassSimplifyStatements.cpp

```diff
@@ -0,0 +1,106 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
+//                          _
+//                         | |
+//                       __| | __ ___      ___ ___
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+
+#include "PassSimplifyStatements.h"
+#include "dawn/AST/ASTExpr.h"
+#include "dawn/IIR/ASTFwd.h"
+#include "dawn/IIR/AccessComputation.h"
+#include "dawn/IIR/DoMethod.h"
+#include "dawn/IIR/StencilInstantiation.h"
+#include "dawn/Support/Type.h"
+#include <memory>
+
+namespace dawn {
+
+namespace {
+class IncrementDecrementReplacer : public ast::ASTVisitorPostOrder {
+  std::vector<std::shared_ptr<ast::Stmt>> statements_;
+
+public:
+  std::shared_ptr<iir::Expr>
+  postVisitNode(std::shared_ptr<iir::UnaryOperator> const& unaryOp) override {
+    auto sourceLoc = unaryOp->getSourceLocation();
+    std::shared_ptr<ast::BinaryOperator> binOp;
+    if(unaryOp->getOp() == "++") {
+      binOp = std::make_shared<ast::BinaryOperator>(
+          unaryOp->getOperand()->clone(), "+",
+          std::make_shared<ast::LiteralAccessExpr>("1", BuiltinTypeID::Integer, sourceLoc),
+          sourceLoc);
+    } else if(unaryOp->getOp() == "--") {
+      binOp = std::make_shared<ast::BinaryOperator>(
+          unaryOp->getOperand()->clone(), "-",
+          std::make_shared<ast::LiteralAccessExpr>("1", BuiltinTypeID::Integer, sourceLoc),
+          sourceLoc);
```

Comment:
Doesn't work in this case:

```
if (i == 0 && ++i == 1) { /* ... */ }
```
will become:
```
i = i + 1;
if (i == 0 && i == 1) { /* ... */ }
```

Reply:
This is an interesting one! Can we resolve this (without introducing a temporary variable)?

Reply:
A valid transformation output would be:
```
if (i == 0) {
  i = i + 1;
  if (i == 1) { /* ... */ }
}
```
But I don't think that such an output would be suitable for a general transformation algorithm.

Note: this isn't equivalent if SIR also has C++'s short-circuit semantics (it would also introduce more ops):
```
i = i + 1;
if (i - 1 == 0 && i == 1) { /* ... */ }
```

Reply:
As this is not an easy problem to solve, let's leave this open for the next syntax workshop.

Reply:
I agree. But do we then merge a pass that only works for a subset of SIR and could change semantics if SIR outside that subset is given?

Reply:
@BenWeber42 I wouldn't merge it for now. Let's briefly present the problem at the review meeting. A quick solution could be saying that increment and decrement ops can't be nested in expressions (i.e. only allowed in statements like i++; i--;) from the beginning (frontends/SIR). If this solution is not widely accepted, then this problem will need to be discussed more in detail, because it seems that the only way to solve it is to do the heavy transformation that you showed.